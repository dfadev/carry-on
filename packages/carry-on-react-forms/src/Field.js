/** @format **/
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/jsx-curly-newline */
import React from "react";
import { State } from "carry-on-react";
import { getIn } from "carry-on-utils";
import FormContext from "./FormContext";

function getVal({ target: { type, value, checked } }) {
  if (type === "number" || type === "range") {
    const parsed = parseFloat(value);
    return Number.isNaN(parsed) ? "" : parsed;
  }

  if (type === "checkbox" || type === "radio") return checked;

  return value;
}

export default ({
  store: propStore,
  form: propForm = "form",
  path = "",
  default: def,
  children = () => null,
  type,
  ...rest
}) => (
  <FormContext.Consumer>
    {({ store, form } = { store: propStore, form: propForm }) => (
      <State path={form} from={store} {...rest} strict>
        {({
          values,
          touched,
          errors,
          visited,
          setFieldValue,
          hasVisited,
          isTouched,
          setFieldVisited,
          setFieldTouched,
          setFieldError
        } = {}) =>
          children({
            touched: getIn(touched, path, false),
            error: getIn(errors, path, undefined),
            visited: getIn(visited, path, false),
            element: {
              onFocus: () => !hasVisited(path) && setFieldVisited(path, true),
              onChange: e => setFieldValue(path, getVal(e)),
              onBlur: () => !isTouched(path) && setFieldTouched(path, true),
              [type === "checkbox" || type === "radio"
                ? "checked"
                : "value"]: getIn(values, path, def)
            },
            setValue: val => setFieldValue(path, val),
            setVisited: val => setFieldVisited(path, val),
            setTouched: val => setFieldTouched(path, val),
            setError: val => setFieldError(path, val)
          })
        }
      </State>
    )}
  </FormContext.Consumer>
);
