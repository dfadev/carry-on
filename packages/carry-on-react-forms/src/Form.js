/** @format **/
import React from "react";
import { State } from "carry-on-react";

export default ({ store, form = "form", children, ...rest }) => (
  <State path={form} from={store} {...rest}>
    {state => (
      <form onSubmit={state.submit} onReset={state.reset}>
        {children}
      </form>
    )}
  </State>
);
