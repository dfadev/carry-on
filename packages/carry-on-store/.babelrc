{
  "presets": [
    ["env", {
      "modules": false,
      "targets": { "browsers": ["last 2 versions"] }
    }],
    "stage-2",
    "optimized"
  ],
  "env": {
    "development": {
      "plugins": ["external-helpers"],
    },
    "production": {
      "plugins": ["external-helpers"],
    },
    "test": {
      "presets": ["env", "stage-2"]
    }
  }
}
