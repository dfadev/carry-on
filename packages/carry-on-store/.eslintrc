{
  "parser": "babel-eslint",
  "extends": ["airbnb-base", "prettier"],
  "env": { "es6": true, "browser": true },
  "rules": {
    "no-empty-pattern": "off",
    "no-debugger": "warn",
    "no-unused-expressions": "off",
    "no-unused-vars": "warn",
    "no-plusplus": "off",
    "import/no-unresolved": "off",
    "no-restricted-syntax": "off",
    "no-underscore-dangle": "off",
    "prefer-template": "off",
    "spaced-comment": "off",
    "no-param-reassign": "off",
    "no-return-assign": "off",
    "no-multi-assign": "off",
    "one-var": "off",
    "prefer-destructuring": "off",
  }
}
