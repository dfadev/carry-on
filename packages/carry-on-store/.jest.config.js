module.exports = {
  //setupTestFrameworkScriptFile: "./node_modules/jest-enzyme/lib/index.js",
  setupTestFrameworkScriptFile: "./test/jest.setup.js",
  verbose: true,
  testPathIgnorePatterns: ["<rootDir>/build/", "<rootDir>/node_modules/", "<rootDir>/examples/", "<rootDir>/lib", "<rootDir>/src/jest.setup.js"],
  collectCoverage: true,
  coverageDirectory: "<rootDir>/coverage/",
  moduleNameMapper: {
    "^carry-on-utils$":
      "<rootDir>/node_modules/carry-on-utils/dist/carry-on-utils.esm.js",
  }
};
