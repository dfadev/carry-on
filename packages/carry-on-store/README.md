# carry-on-store

# STATUS: WIP SUBJECT TO CHANGE

- This is the core module that provides the store.

Read the documentation at: https://dfadev.github.io/carry-on
