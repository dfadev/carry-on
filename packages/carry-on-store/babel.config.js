module.exports = {
  presets: [
    [
      "@babel/preset-env",
      {
        modules: false,
        targets: {
          browsers: [">3%"]
        }
      }
    ]
  ],
  env: {
    test: {
      plugins: [
        "@babel/plugin-transform-modules-commonjs",
        [
          "@babel/plugin-transform-runtime",
          {
            regenerator: true
          }
        ]
      ],
      presets: ["@babel/preset-env"]
    }
  }
};
