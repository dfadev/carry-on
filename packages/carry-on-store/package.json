{
  "name": "carry-on-store",
  "version": "0.0.50",
  "description": "State manager",
  "source": "./src/index.js",
  "module": "./dist/carry-on-store.esm.js",
  "main": "./dist/carry-on-store.cjs.js",
  "browser": "./dist/carry-on-store.umd.js",
  "scripts": {
    "clean": "rimraf lib dist es coverage",
    "format": "prettier --write \"{src,test}/**/*.js\"",
    "format:check": "prettier --list-different \"{src,test}/**/*.js\"",
    "lint": "eslint src test",
    "pretest": "npm run build",
    "test": "cross-env BABEL_ENV=test jest --config ./.jest.config.js",
    "test:watch": "npm test -- --watchAll",
    "test:cov": "npm test -- --coverage",
    "build": "rollup -c rollup.config.js",
    "build-dev": "rollup -c rollup.config.js --configDebug",
    "watch": "npm run build -- -w"
  },
  "author": "Russ Panula <russ@dfa.net>",
  "license": "MIT",
  "devDependencies": {
    "@babel/core": "^7.7.2",
    "@babel/plugin-transform-modules-commonjs": "^7.7.0",
    "@babel/plugin-transform-runtime": "^7.6.2",
    "@babel/preset-env": "^7.7.1",
    "@testing-library/jest-dom": "^4.2.4",
    "babel-eslint": "^10.0.3",
    "babel-jest": "^24.9.0",
    "cross-env": "^6.0.3",
    "eslint": "^6.6.0",
    "eslint-config-airbnb-base": "^14.0.0",
    "eslint-config-prettier": "^6.6.0",
    "eslint-plugin-import": "^2.18.2",
    "jest": "^24.9.0",
    "prettier": "^1.19.1",
    "rimraf": "^3.0.0",
    "rollup": "^1.27.0",
    "rollup-plugin-babel": "^4.3.3",
    "rollup-plugin-eslint": "^7.0.0",
    "snapshot-diff": "^0.6.1"
  },
  "sideEffects": false,
  "gitHead": "e21942a1d6e23d964f9b74bd1c9c3744cb7b3c18",
  "dependencies": {
    "@babel/runtime": "^7.7.2",
    "carry-on-utils": "^0.0.50",
    "immer": "^5.0.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dfadev/carry-on.git"
  }
}
