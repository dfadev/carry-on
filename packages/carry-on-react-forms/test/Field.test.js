/**
 * @jest-environment jsdom
 */
import React from "react";
import { render } from "@testing-library/react";
import { get, initStores } from "carry-on-store";
import Form from "../src/Form";
import Field from "../src/Field";

beforeEach(() => initStores());

it("should render <Field>", () => {
  expect(render(<Field />).asFragment()).toMatchSnapshot();
});

const formState = (formId = "form") => ({
  state: ({ get, set }) => ({
    [formId]: {
      initialValues: {
        field1: "value1",
        field2: "value2",
        checkbox: false
      },
      onSubmit: vals => {},
      onValidate: undefined,
      onReset: vals => {}
    }
  })
});

it("should render a simple input field", () => {
  const { asFragment } = render(
    <Form register={formState()}>
      <Field path="field1">
        {state => {
          expect(state).toMatchSnapshot("prop match");
          return <input {...state.element} />;
        }}
      </Field>
    </Form>
  );

  expect(asFragment()).toMatchSnapshot("render match");
  expect(get()).toMatchSnapshot("store match");
});

it("should render a simple input field in named form", () => {
  const { asFragment } = render(
    <Form id="theform" register={formState("theform")}>
      <Field path="field1">
        {state => {
          expect(state).toMatchSnapshot("prop match");
          return <input {...state.element} />;
        }}
      </Field>
    </Form>
  );

  expect(asFragment()).toMatchSnapshot("render match");
  expect(get()).toMatchSnapshot("store match");
});
