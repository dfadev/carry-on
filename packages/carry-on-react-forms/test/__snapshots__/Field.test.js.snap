// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <Field> 1`] = `<DocumentFragment />`;

exports[`should render a simple input field in named form: prop match 1`] = `
Object {
  "element": Object {
    "onBlur": [Function],
    "onChange": [Function],
    "onFocus": [Function],
    "value": "value1",
  },
  "error": undefined,
  "setError": [Function],
  "setTouched": [Function],
  "setValue": [Function],
  "setVisited": [Function],
  "touched": false,
  "visited": false,
}
`;

exports[`should render a simple input field in named form: render match 1`] = `
<DocumentFragment>
  <form>
    <input
      value="value1"
    />
  </form>
</DocumentFragment>
`;

exports[`should render a simple input field in named form: store match 1`] = `
Object {
  "theform": Object {
    "errors": Object {},
    "hasError": [Function],
    "hasVisited": [Function],
    "initialValues": Object {
      "checkbox": false,
      "field1": "value1",
      "field2": "value2",
    },
    "isPristine": true,
    "isSubmitting": false,
    "isTouched": [Function],
    "isValid": true,
    "isValidating": false,
    "onReset": [Function],
    "onSubmit": [Function],
    "onValidate": undefined,
    "origState": Object {
      "errors": Object {},
      "hasError": [Function],
      "hasVisited": [Function],
      "isPristine": true,
      "isSubmitting": false,
      "isTouched": [Function],
      "isValid": true,
      "isValidating": false,
      "reset": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFieldVisited": [Function],
      "setInitialValues": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "submit": [Function],
      "touched": Object {},
      "validate": [Function],
      "validation": undefined,
      "values": Object {
        "checkbox": false,
        "field1": "value1",
        "field2": "value2",
      },
      "visited": Object {},
    },
    "reset": [Function],
    "setErrors": [Function],
    "setFieldError": [Function],
    "setFieldTouched": [Function],
    "setFieldValue": [Function],
    "setFieldVisited": [Function],
    "setInitialValues": [Function],
    "setTouched": [Function],
    "setValues": [Function],
    "submit": [Function],
    "touched": Object {},
    "validate": [Function],
    "values": Object {
      "checkbox": false,
      "field1": "value1",
      "field2": "value2",
    },
    "visited": Object {},
  },
}
`;

exports[`should render a simple input field: prop match 1`] = `
Object {
  "element": Object {
    "onBlur": [Function],
    "onChange": [Function],
    "onFocus": [Function],
    "value": "value1",
  },
  "error": undefined,
  "setError": [Function],
  "setTouched": [Function],
  "setValue": [Function],
  "setVisited": [Function],
  "touched": false,
  "visited": false,
}
`;

exports[`should render a simple input field: render match 1`] = `
<DocumentFragment>
  <form>
    <input
      value="value1"
    />
  </form>
</DocumentFragment>
`;

exports[`should render a simple input field: store match 1`] = `
Object {
  "form": Object {
    "errors": Object {},
    "hasError": [Function],
    "hasVisited": [Function],
    "initialValues": Object {
      "checkbox": false,
      "field1": "value1",
      "field2": "value2",
    },
    "isPristine": true,
    "isSubmitting": false,
    "isTouched": [Function],
    "isValid": true,
    "isValidating": false,
    "onReset": [Function],
    "onSubmit": [Function],
    "onValidate": undefined,
    "origState": Object {
      "errors": Object {},
      "hasError": [Function],
      "hasVisited": [Function],
      "isPristine": true,
      "isSubmitting": false,
      "isTouched": [Function],
      "isValid": true,
      "isValidating": false,
      "reset": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFieldVisited": [Function],
      "setInitialValues": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "submit": [Function],
      "touched": Object {},
      "validate": [Function],
      "validation": undefined,
      "values": Object {
        "checkbox": false,
        "field1": "value1",
        "field2": "value2",
      },
      "visited": Object {},
    },
    "reset": [Function],
    "setErrors": [Function],
    "setFieldError": [Function],
    "setFieldTouched": [Function],
    "setFieldValue": [Function],
    "setFieldVisited": [Function],
    "setInitialValues": [Function],
    "setTouched": [Function],
    "setValues": [Function],
    "submit": [Function],
    "touched": Object {},
    "validate": [Function],
    "values": Object {
      "checkbox": false,
      "field1": "value1",
      "field2": "value2",
    },
    "visited": Object {},
  },
}
`;
